x/slashing/handler.go

```diff
@@ -0,0 +1,58 @@
+package slashing
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func NewHandler(k Keeper) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+		// NOTE msg already has validate basic run
+		switch msg := msg.(type) {
+		case MsgUnrevoke:
+			return handleMsgUnrevoke(ctx, msg, k)
+		default:
+			return sdk.ErrTxDecode("invalid message parse in staking module").Result()
+		}
+	}
+}
+
+// Validators must submit a transaction to unrevoke themselves after
+// having been revoked (and thus unbonded) for downtime
+func handleMsgUnrevoke(ctx sdk.Context, msg MsgUnrevoke, k Keeper) sdk.Result {
+
+	// Validator must exist
+	validator := k.stakeKeeper.Validator(ctx, msg.ValidatorAddr)
+	if validator == nil {
+		return ErrNoValidatorForAddress(k.codespace).Result()
+	}
+
+	addr := validator.GetPubKey().Address()
+
+	// Signing info must exist
+	info, found := k.getValidatorSigningInfo(ctx, addr)
+	if !found {
+		return ErrNoValidatorForAddress(k.codespace).Result()
+	}
+
+	// Cannot be unrevoked until out of jail
+	if ctx.BlockHeader().Time < info.JailedUntil {
+		return ErrValidatorJailed(k.codespace).Result()
+	}
```

Comment:
As I mention later on - I think this logic of if the unrevoke is valid should be moved into `keeper` thoughts?

Reply:
no  you're totally correct - I just don't like Unrevoke being generically exposed - maybe not an issue though should just audit the code. OK sounds fine as is